lib/data/external/AwsClient.js

```diff
@@ -43,23 +73,33 @@ class AwsClient {
         }
 
         if (!stream) {
-            return this._client.putObject(uploadParams, err => {
+            return this._client.putObject(uploadParams, (err, data) => {
                 if (err) {
-                    const log = createLogger(reqUids);
                     logHelper(log, 'error', 'err from data backend',
                       err, this._dataStoreName);
                     return callback(errors.InternalError
                       .customizeDescription('Error returned from ' +
                       `AWS: ${err.message}`)
                     );
                 }
-                return callback(null, awsKey);
+                if (!data.VersionId) {
+                    const error = errors.InternalError.customizeDescription(
+                        'Invalid state. Please ensure versioning is enabled ' +
+                        'in AWS for the location constraint\'s bucket and ' +
+                        'try again.'
+                    );
+                    logHelper(log, 'error', 'missing version id for data ' +
+                    'backend object', error, this._dataStoreName);
+                    return callback(error);
+                }
+                const params = { dataStoreVersionId: data.VersionId };
```

Comment:
do you foresee us having more params? if not, can we just send the versionId back?